osquery/utils/windows/lzxpress.cpp

```diff
@@ -0,0 +1,105 @@
+/**
+ * Copyright (c) 2014-present, The osquery authors
+ *
+ * This source code is licensed as defined by the LICENSE file found in the
+ * root directory of this source tree.
+ *
+ * SPDX-License-Identifier: (Apache-2.0 OR GPL-2.0-only)
+ */
+
+#include <osquery/logger/logger.h>
+#include <osquery/utils/windows/lzxpress.h>
+
+#include <boost/algorithm/hex.hpp>
+#include <boost/algorithm/string.hpp>
+
+namespace osquery {
+
+std::string decompressLZxpress(std::vector<char> prefetch_data,
+                               unsigned long size) {
+  static HMODULE hDLL =
+      LoadLibraryExW(L"ntdll.dll", NULL, LOAD_LIBRARY_SEARCH_SYSTEM32);
+  if (hDLL == nullptr) {
+    LOG(ERROR) << "Failed to load ntdll.dll";
+    return "Error";
+  }
+
+  typedef HRESULT(WINAPI * pRtlDecompressBufferEx)(
+      _In_ unsigned short format,
+      _Out_ unsigned char* uncompressedBuffer,
+      _In_ unsigned long uncompressedSize,
+      _In_ unsigned char* data,
+      _In_ unsigned long dataSize,
+      _Out_ unsigned long* finalSize,
+      _In_ PVOID workspace);
+
+  typedef HRESULT(WINAPI * pRtlGetCompressionWorkSpaceSize)(
+      _In_ unsigned short format,
+      _Out_ unsigned long* bufferWorkSpaceSize,
+      _Out_ unsigned long* fragmentWorkSpaceSize);
+
+  pRtlDecompressBufferEx RtlDecompressBufferEx;
+  pRtlGetCompressionWorkSpaceSize RtlGetCompressionWorkSpaceSize;
+  RtlGetCompressionWorkSpaceSize =
+      (pRtlGetCompressionWorkSpaceSize)GetProcAddress(
+          hDLL, "RtlGetCompressionWorkSpaceSize");
+  if (RtlGetCompressionWorkSpaceSize == nullptr) {
+    LOG(ERROR) << "Failed to load function RtlGetCompressionWorkSpaceSize";
+    return "Error";
+  }
+  RtlDecompressBufferEx =
+      (pRtlDecompressBufferEx)GetProcAddress(hDLL, "RtlDecompressBufferEx");
+
+  // Check for decompression function, only exists on Win8+
+  if (RtlDecompressBufferEx == nullptr) {
+    LOG(ERROR) << "Failed to load function RtlDecompressBufferEx";
+    return "Error";
+  }
+  unsigned long bufferWorkSpaceSize = 0ul;
+  unsigned long fragmentWorkSpaceSize = 0ul;
+  auto results = RtlGetCompressionWorkSpaceSize(COMPRESSION_FORMAT_XPRESS_HUFF,
+                                                &bufferWorkSpaceSize,
+                                                &fragmentWorkSpaceSize);
+  if (results != 0) {
+    LOG(ERROR) << "Failed to set compression workspace size";
+    return "Error";
+  }
+  unsigned char* compressed_data = new unsigned char[prefetch_data.size() - 8];
```

Comment:
Is it possible to use an `std::vector<>` instead?

Reply:
These variables are passed to the Windows api functions used to decompress the prefetch files.  
The functions dont accept vectors as an arg?

Reply:
I think what @alessandrogario is trying to imply is this:

```
std::vector<BYTE> compressed_data;
compressed_data.resize(prefetch_data.size()-8;
...
RtlDecompressBufferEx(..., compressed_data.data() ...
```

By the way, I recommend using Windows API data types vs primitive data types, e.g., `BYTE` vs `unsigned char`. This is for  portability. So, if a Windows API function expects `BYTE` or returns a `DWORD`, passing it `unsigned char` or receiving in an `unsigned long` will introduce subtle bugs while making code less portable.

So, I would recommend updating to what these functions actually expect as per documentation.

Reply:
Thanks for the clarification @farfella  
I was originally going to use Windows API data types but when I reviewed the old prefetch table attempt, there was a comment to avoid windows types?
https://github.com/MobeenAshraf/osquery/commit/c946f3c75bbfd0676db585c8157bf04f539cf1ca#r21629430

The comment is pretty old, so maybe its fine now?
I can change to the Windows API data types, if that preferred now

Reply:
I agree with @farfella, it is best to use the types the win32 API expect so we don't run into incompatibility issues

Reply:
sounds good, switched to win32 types